core/variables.js

```diff
@@ -340,9 +351,21 @@ Blockly.Variables.renameVariable = function(workspace, variable,
     Blockly.Variables.promptName(promptText, defaultName,
         function(newName) {
           if (newName) {
-            workspace.renameVariableById(variable.getId(), newName);
-            if (opt_callback) {
-              opt_callback(newName);
+            var existing = Blockly.Variables.nameUsedWithOtherType_(newName,
+                variable.type, workspace);
+            if (existing) {
+              var msg =
+                  Blockly.Msg.VARIABLE_ALREADY_EXISTS_FOR_ANOTHER_TYPE.replace(
+                  '%1', newName.toLowerCase()).replace('%2', existing.type);
+              Blockly.alert(msg,
+                function() {
+                  promptAndCheckWithAlert(newName);  // Recurse
+                });
+            } else {
+              workspace.renameVariableById(variable.getId(), newName);
+              if (opt_callback) {
+                opt_callback(newName);
+              }
             }
           } else {
             // User canceled prompt without a value.
```

Comment:
Just `// User canceled prompt.`
(Same above, but I can't comment there.)

Reply:
Done for both.